packages/open-next/src/build/middleware/buildNodeMiddleware.ts

```diff
@@ -0,0 +1,138 @@
+import * as buildHelper from "../helper.js";
+import { installDependencies } from "../installDeps.js";
+
+type Override = OverrideOptions & {
+  originResolver?: LazyLoadedOverride<OriginResolver> | IncludedOriginResolver;
+};
+
+export async function buildExternalNodeMiddleware(
+  options: buildHelper.BuildOptions,
+) {
+  const { appBuildOutputPath, config, outputDir } = options;
+  if (!config.middleware?.external) {
+    throw new Error(
+      "This function should only be called for external middleware",
+    );
+  }
+  const outputPath = path.join(outputDir, "middleware");
+  fs.mkdirSync(outputPath, { recursive: true });
+
+  // Copy open-next.config.mjs
+  buildHelper.copyOpenNextConfig(
+    options.buildDir,
+    outputPath,
+    await buildHelper.isEdgeRuntime(config.middleware.override),
+  );
+  const overrides = {
+    ...config.middleware.override,
+    originResolver: config.middleware.originResolver,
+  };
+  const includeCache = config.dangerous?.enableCacheInterception;
+  const packagePath = buildHelper.getPackagePath(options);
+
+  // TODO: change this so that we don't copy unnecessary files
+  await copyTracedFiles(
+    appBuildOutputPath,
+    packagePath,
+    outputPath,
+    [],
+    false,
+    true,
+  );
+
+  function override<T extends keyof Override>(target: T) {
+    return typeof overrides?.[target] === "string"
+      ? overrides[target]
+      : undefined;
+  }
+
+  // Bundle middleware
+  await buildHelper.esbuildAsync(
+    {
+      entryPoints: [
+        path.join(options.openNextDistDir, "adapters", "middleware.js"),
+      ],
+      outfile: path.join(outputPath, "handler.mjs"),
+      external: ["./.next/*"],
+      platform: "node",
+      plugins: [
+        openNextResolvePlugin({
+          overrides: {
+            wrapper: override("wrapper") ?? "aws-lambda",
+            converter: override("converter") ?? "aws-cloudfront",
+            ...(includeCache
+              ? {
+                  tagCache: override("tagCache") ?? "dynamodb-lite",
+                  incrementalCache: override("incrementalCache") ?? "s3-lite",
+                  queue: override("queue") ?? "sqs-lite",
+                }
+              : {}),
+            originResolver: override("originResolver") ?? "pattern-env",
+            proxyExternalRequest: override("proxyExternalRequest") ?? "node",
+          },
+          fnName: "middleware",
+        }),
+        openNextReplacementPlugin({
+          name: "externalMiddlewareOverrides",
+          target: getCrossPlatformPathRegex("adapters/middleware.js"),
+          deletes: includeCache ? [] : ["includeCacheInMiddleware"],
+        }),
+        openNextExternalMiddlewarePlugin(
+          path.join(
+            options.openNextDistDir,
+            "core",
+            "nodeMiddlewareHandler.js",
+          ),
+        ),
+      ],
+      banner: {
+        js: [
+          `globalThis.monorepoPackagePath = "${packagePath}";`,
```

Comment:
tiny nit: single quote around the package name like the format of the imports below.